### PR TITLE
Allow injection of helper bytecode as resources

### PR DESCRIPTION
**Pull request by JonasKunz**
Allows to inject the bytecode of proxies as resources. This is required for example for injecting spring-boot autoconfigurations.

### CODE REVIEW

**Comment by JonasKunz** [comment]
@open-telemetry/java-instrumentation-maintainers Please add the `test Indy` label to this PR. I'll then enable the spring-actuator instrumentation in this PR to run with indy.

**Comment by trask** [comment]
> @open-telemetry/java-instrumentation-maintainers Please add the `test Indy` label to this PR. I'll then enable the spring-actuator instrumentation in this PR to run with indy.

done 👍

File: muzzle/src/main/java/io/opentelemetry/javaagent/tooling/ByteArrayUrl.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+import java.security.PrivilegedAction;
+
+public class ByteArrayUrl {
+
+  private ByteArrayUrl() {}
+
+  private static final String URL_SCHEMA = "x-otel-binary";
+
+  @SuppressWarnings("removal")
+  public static URL create(String uniqueName, byte[] data) {
```
Comment (laurit) [inline muzzle/src/main/java/io/opentelemetry/javaagent/tooling/ByteArrayUrl.java:27]: I don't get why these would need to have a unique name

File: muzzle/src/main/java/io/opentelemetry/javaagent/tooling/ByteArrayUrl.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+              });
+    } else {
+      return doCreate(uniqueName, data);
+    }
+  }
+
+  private static URL doCreate(String uniqueName, byte[] data) {
+    try {
+      String hostname = URLEncoder.encode(uniqueName, StandardCharsets.UTF_8.toString());
+      return new URL(URL_SCHEMA, hostname, -1, "", new ByteArrayUrlStreamHandler(data));
```
Comment (laurit) [inline muzzle/src/main/java/io/opentelemetry/javaagent/tooling/ByteArrayUrl.java:42]: Using the hostname component instead of file feels a bit weird. Imo using the file would be better, at least then it wouldn't try to do a dns lookup if somehow hashCode or equals is called on this url.

**Review by laurit — APPROVED** [review]

File: muzzle/src/main/java/io/opentelemetry/javaagent/tooling/ByteArrayUrl.java (hunk 1 above)
Comment (JonasKunz) [reply muzzle/src/main/java/io/opentelemetry/javaagent/tooling/ByteArrayUrl.java:27]: This was just done based on what ByteBuddy does. I assume it is just a nice-to-have so that `toString()` returns something which allows users to identify the content when debugging.

File: muzzle/src/main/java/io/opentelemetry/javaagent/tooling/ByteArrayUrl.java (hunk 2 above)
Comment (JonasKunz) [reply muzzle/src/main/java/io/opentelemetry/javaagent/tooling/ByteArrayUrl.java:42]: This was also simply copied from ByteBuddy. I didn't alter it because it has already been battle-tested there.
If you want to, we can however switch it to the file and see what happens.

File: muzzle/src/main/java/io/opentelemetry/javaagent/tooling/ByteArrayUrl.java (hunk 1 above)
Comment (laurit) [reply muzzle/src/main/java/io/opentelemetry/javaagent/tooling/ByteArrayUrl.java:27]: Let me rephrase. Why is the parameter called `uniqueName`? Does it really have to be unique?

Comment (JonasKunz) [reply muzzle/src/main/java/io/opentelemetry/javaagent/tooling/ByteArrayUrl.java:27]: Ah I see. No it doesn't have to be unique. I'll change it to `contentName`.

File: muzzle/src/main/java/io/opentelemetry/javaagent/tooling/ByteArrayUrl.java (hunk 2 above)
Comment (laurit) [reply muzzle/src/main/java/io/opentelemetry/javaagent/tooling/ByteArrayUrl.java:42]: I still believe that it is better to use the file component rather than the host, it is what everybody else is doing.
